File: tools/dump.py
"""Python 版の純関数の出力を、そのまま比較できる形で吐く。tools/dump.mjs と対にする。"""

import hashlib
import sys
from datetime import date, datetime, timedelta

from app import config, meals, slots

TZ = config.TZ
FIRST = date(2026, 8, 16)
LAST = date(2026, 8, 29)


def days():
    day = FIRST
    while day <= LAST:
        yield day
        day += timedelta(days=1)


def grid(step_minutes):
    now = datetime(2026, 8, 16, 0, 0, tzinfo=TZ)
    end = datetime(2026, 8, 29, 0, 0, tzinfo=TZ)
    while now <= end:
        yield now
        now += timedelta(minutes=step_minutes)


def emit(out, label, value):
    out.write(f"{label}\t{value}\n")


def main():
    out = sys.stdout

    for section in config.SECTIONS:
        for room in config.ROOMS:
            emit(out, f"slot_starts {section} {room}", ",".join(slots.slot_starts(section, room)))

    for day in days():
        emit(out, f"sections_for_date {day}", ",".join(slots.sections_for_date(day)))
        emit(out, f"format_day {day}", slots.format_day(day))

    for meal in meals.all_meals():
        emit(out, f"meal {meal.day} {meal.kind}", meals.format_meal(meal))
        emit(out, f"deadline {meal.day} {meal.kind}", meals.vote_deadline(meal).isoformat())

    for day in days():
        for section in config.SECTIONS:
            for room in config.ROOMS:
                for slot in slots.slot_starts(section, room):
                    when = slots.slot_datetime(day, slot)
                    emit(out, f"slot_datetime {day} {slot}", when.isoformat())

    # 1 分刻み。滞在日の境目を隅々まで見る。
    for now in grid(1):
        stay = slots.stay_date(now)
        selectable = ",".join(str(d) for d in slots.selectable_dates(now))
        emit(out, f"stay {now.isoformat()}", f"{stay}|{selectable}")

    # 30 分刻み。全ての日・区分・浴室・枠についての可否をまとめて digest にする。
    for now in grid(30):
        parts = []
        for day in days():
            for section in config.SECTIONS:
                for room in config.ROOMS:
                    for slot in slots.slot_starts(section, room):
                        parts.append(slots.check_reservable(day, section, room, slot, now))
            parts.append("1" if slots.has_open_slot(day, now) else "0")
        for meal in meals.all_meals():
            parts.append(meals.vote_state(meal, now))
        parts.append(",".join(f"{m.day}{m.kind}" for m in meals.visible_meals(now)))
        joined = "|".join(parts)
        emit(out, f"digest {now.isoformat()}", hashlib.sha1(joined.encode()).hexdigest())


main()

File: tools/dump.mjs
/** Node 版の純関数の出力を、tools/dump.py と同じ形で吐く。 */

import { createHash } from "node:crypto";

import * as config from "../src/config.js";
import * as jst from "../src/jst.js";
import * as meals from "../src/meals.js";
import * as slots from "../src/slots.js";

const FIRST = "2026-08-16";
const LAST = "2026-08-29";

function days() {
  const out = [];
  for (let day = FIRST; day <= LAST; day = jst.addDays(day, 1)) out.push(day);
  return out;
}

function grid(stepMinutes) {
  const out = [];
  const end = jst.instant("2026-08-29", 0);
  for (let now = jst.instant("2026-08-16", 0); now <= end; now += stepMinutes * 60_000) {
    out.push(now);
  }
  return out;
}

const lines = [];
const emit = (label, value) => lines.push(`${label}\t${value}`);

for (const section of config.SECTIONS) {
  for (const room of config.ROOMS) {
    emit(`slot_starts ${section} ${room}`, slots.slotStarts(section, room).join(","));
  }
}

for (const day of days()) {
  emit(`sections_for_date ${day}`, slots.sectionsForDate(day).join(","));
  emit(`format_day ${day}`, slots.formatDay(day));
}

for (const meal of meals.allMeals()) {
  emit(`meal ${meal.day} ${meal.kind}`, meals.formatMeal(meal));
  emit(`deadline ${meal.day} ${meal.kind}`, jst.iso(meals.voteDeadline(meal)));
}

for (const day of days()) {
  for (const section of config.SECTIONS) {
    for (const room of config.ROOMS) {
      for (const slot of slots.slotStarts(section, room)) {
        emit(`slot_datetime ${day} ${slot}`, jst.iso(slots.slotInstant(day, slot)));
      }
    }
  }
}

for (const now of grid(1)) {
  emit(`stay ${jst.iso(now)}`, `${slots.stayDate(now)}|${slots.selectableDates(now).join(",")}`);
}

for (const now of grid(30)) {
  const parts = [];
  for (const day of days()) {
    for (const section of config.SECTIONS) {
      for (const room of config.ROOMS) {
        for (const slot of slots.slotStarts(section, room)) {
          parts.push(slots.checkReservable(day, section, room, slot, now));
        }
      }
    }
    parts.push(slots.hasOpenSlot(day, now) ? "1" : "0");
  }
  for (const meal of meals.allMeals()) parts.push(meals.voteState(meal, now));
  parts.push(meals.visibleMeals(now).map((m) => `${m.day}${m.kind}`).join(","));
  emit(`digest ${jst.iso(now)}`, createHash("sha1").update(parts.join("|")).digest("hex"));
}

process.stdout.write(`${lines.join("\n")}\n`);
